test/integration-tests/Uniswap.test.ts

```diff
@@ -1196,6 +1206,480 @@ describe('Uniswap V2 and V3 Tests:', () => {
+          amount0Min: 0,
+          amount1Min: 0,
+          deadline: MAX_UINT,
+        }
+
+        const encodedDecreaseCall = encodeDecreaseLiquidity(decreaseParams)
+
+        const COLLECT_STRUCT = '(uint256 tokenId,address recipient,uint256 amount0Max,uint256 amount1Max)'
+        const collectParams = {
+          tokenId: tokenId,
+          recipient: bob.address,
+          amount0Max: MAX_UINT128,
+          amount1Max: MAX_UINT128,
+        }
+
+        const abi = new ethers.utils.AbiCoder()
+        const encodedCollectParams = abi.encode([COLLECT_STRUCT], [collectParams])
+        const functionSignatureCollect = ethers.utils.id('collect((uint256,address,uint128))').substring(0, 10)
+        const encodedCollectCall = functionSignatureCollect + encodedCollectParams.substring(2)
+
+        planner.addCommand(CommandType.V3_POSM_CALL, [encodedDecreaseCall])
+        planner.addCommand(CommandType.V3_POSM_CALL, [encodedCollectCall])
+
+        await expect(executeRouter(planner)).to.be.revertedWithCustomError(router, 'InvalidV3Action')
+      })
+
+      it('cannot call collect with improper params', async () => {
+        // first we need to permit the router to spend the nft
+        let { v, r, s } = await getPermitNFTSignature(bob, v3NFTPositionManager, router.address, tokenId, MAX_UINT)
+        planner.addCommand(CommandType.ERC721_PERMIT, [ADDRESS_THIS, tokenId, MAX_UINT, v, r, s])
+
+        let position = await v3NFTPositionManager.positions(tokenId)
+        let liquidity = position.liquidity
+
+        const decreaseParams = {
+          tokenId: tokenId,
+          liquidity: liquidity,
+          amount0Min: 0,
+          amount1Min: 0,
+          deadline: MAX_UINT,
+        }
+
+        const encodedDecreaseCall = encodeDecreaseLiquidity(decreaseParams)
+
+        const COLLECT_STRUCT = '(uint256 tokenId,address recipient,uint256 amount0Max)'
+        const collectParams = { tokenId: tokenId, recipient: bob.address, amount0Max: MAX_UINT128 }
+
+        const abi = new ethers.utils.AbiCoder()
+        const encodedCollectParams = abi.encode([COLLECT_STRUCT], [collectParams])
+        const functionSignatureCollect = ethers.utils.id('collect((uint256,address,uint128,uint128))').substring(0, 10)
+        const encodedCollectCall = functionSignatureCollect + encodedCollectParams.substring(2)
+
+        planner.addCommand(CommandType.V3_POSM_CALL, [encodedDecreaseCall])
+        planner.addCommand(CommandType.V3_POSM_CALL, [encodedCollectCall])
+
+        await expect(executeRouter(planner)).to.be.revertedWithCustomError(router, 'CallToV3PositionManagerFailed')
+      })
+
+      it('cannot call collect if not approved', async () => {
+        // first we need to permit the router to spend the nft
+        let { v, r, s } = await getPermitNFTSignature(bob, v3NFTPositionManager, router.address, tokenId, MAX_UINT)
+        planner.addCommand(CommandType.ERC721_PERMIT, [ADDRESS_THIS, tokenId, MAX_UINT, v, r, s])
```

Comment:
would love to see another test where
- bob approves the router, and calls decrease
- separate tx eve tries to call collect to take the tokens